exercises/building_suffix_arrays/descr.md

```diff
@@ -0,0 +1,543 @@
+the letter `'i'`, a bucket of one string that begins with the letter `'m'`,
+a bucket of one string that begins with the letter `'p'`, and a bucket of
+four strings that begin with the letter `'s'`.
+
+In the following, it is necessary for us to know where each bucket begins.
+This information can easily be computed in linear time. This is the topic
+of the next question.
+
+**Question 7.** Define a function `beginning_of_bucket` of type
+`('a -> int) -> 'a array -> bool array`, such that,
+under the assumption that the array `a` is sorted by `key` already,
+the function call `beginning_of_bucket key a`
+produces an array `b` such that `b.(i)` is `true`
+if and only if the index `i` marks the beginning of a new bucket,
+that is, if and only if `i` is zero or the element `a.(i)` does not
+have the same key as its predecessor.
+
+*Example.* If `s` denotes the sorted array of the suffixes of the string
+`"mississipi"`, as shown above, then the expression:
+
+```
+  beginning_of_bucket (fun s -> Char.code s.[0]) s
+```
+
+should produce the following Boolean array:
+
+```
+  [|true; false; false; false; true; true; true; false; false; false|]
+```
+
+## Stages 1, 2, 4, 8...: Jumping from *h* to *2h*
+
+At the beginning of stage *h*, where *h* is at least 1, we have an array *a*
+containing all of the nonempty suffixes of the string *s*.
+Assuming *s* is a string of length *n*, a nonempty
+suffix of *s* is represented as an integer in the semi-open interval
+*[0, n)*. Thus, the array *a*, whose type is `suffix array`, is a
+permutation of *[0, n)*.
+
+At the beginning of stage *h*, we assume that the array *a* is *h*-sorted,
+that is, sorted according to the *h* first characters of each suffix. It is
+useful to think of the array *a* as partitioned into a number of
+*h*-*buckets*, where all of the suffixes inside each bucket are *h*-equal,
+that is, they agree on their *h* first characters.
+
+We assume that a Boolean array *b*, which contains beginning-of-bucket
+marks, is also given.
+
+### Aim
+
+The aim of this stage is to reach a situation where the array *a* is
+*2h*-sorted. This is a more precise preorder: indeed, two suffixes that are
+*2h*-equal must also be *h*-equal. This means that we wish to *refine* the
+current partition. In other words, we wish to *divide* each *h*-bucket into
+one or more *2h*-buckets.
+
+To do so, we must move some suffixes to a different place in the array *a*.
+These moves take place only *within* an *h*-bucket, never from an
+*h*-bucket to another.
+We must also update the array *b* so as to mark the beginning
+of each *2h*-bucket.
+
+It is worth noting that the suffixes in the interval *[n-h+1, n)* will not
+move during this stage. Indeed, these suffixes have fewer than *h* characters,
+so, as far as they are concerned, *h*-order and *2h*-order are the same thing.
+Because *a* is *h*-sorted, each of these suffixes has already reached its
+final position in the array, and inhabits a singleton *h*-bucket.
```

Comment:
```suggestion
final position in the array, and inhabits a singleton *h*-bucket.
```
```suggestion
final position in the array, and inhabits a singleton *h*-bucket in the final array.
```

Reply:
I am not sure about this suggestion. I think each index inhabits a singleton *h*-bucket both in the current array and in the final array. I am not sure if adding "in the final array" clarifies anything. Is something presently unclear?